source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc

```diff
@@ -119,6 +121,20 @@ Network::TransportSocketPtr UpstreamProxyProtocolSocketFactory::createTransportS
                                                        config_.version());
 }
 
+void UpstreamProxyProtocolSocketFactory::hashKey(
+    std::vector<uint8_t>& key, Network::TransportSocketOptionsConstSharedPtr options) const {
+  PassthroughFactory::hashKey(key, options);
+  // Proxy protocol options should only be included in the hash if the upstream
+  // socket intends to use them.
+  if (options) {
+    const auto& proxy_protocol_options = options->proxyProtocolOptions();
```

Comment:
Minor comment: As I was writing this, I realized that we don't have to capture proxy data by copy in the options for proxy protocol. We could just look it up in the filter state here.